openshift/assisted-service
# Bug 1999297: Trigger VIP reallocation only when machine CIDR changes

# Assisted Pull Request

## Description

This PR changes the logic of triggering reallocation of VIPs. Currently
the VIPs are recreated whenever PATCH payload contains MachineNetworks
without checking whether the networks provided are different from the
currently configured ones.

Given that VIPs are always allocated from the first machine network (if
more than one is configured), this logic should be simplified to
triggering the reallocation only if the first machine network is
different than the first machine network already configured.

## List all the issues related to this PR


## What environments does this code impact?


## How was this code tested?




### Manual test

I have created a cluster using test-infra. I configure networking to use the VIP DHCP allocation using the following payload

```
curl --data '{"vip_dhcp_allocation":true,"network_type":"OVNKubernetes","user_managed_networking":false,"cluster_networks":[{"cidr":"10.128.0.0/14","host_prefix":23}],"service_networks":[{"cidr":"172.30.0.0/16"}],"machine_networks":[{"cidr":"192.168.127.0/24"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
```

I wait for the API VIP to be allocated

```
curl http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b 2>/dev/null | jq '.api_vip'
```

Once the VIP appears, I send the following payload (the same `machine_networks` as before)

```
curl --data '{"machine_networks":[{"cidr":"192.168.127.0/24"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
```

by running `watch -n1 curl ...` I see that at no point API VIP was `null`.

I send bigger payload, with explicit VIP DHCP allocation enabled

```
curl --data '{"vip_dhcp_allocation":true,"network_type":"OVNKubernetes","user_managed_networking":false,"cluster_networks":[{"cidr":"10.128.0.0/14","host_prefix":23}],"service_networks":[{"cidr":"172.30.0.0/16"}],"machine_networks":[{"cidr":"192.168.127.0/24"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
```

and again by running `watch -n1 curl ...` I see that at no point API VIP was `null`.

I change the machine network by sending the following payload (the same as the UI sends when changing using the browser)

```
curl --data '{"vip_dhcp_allocation":true,"network_type":"OVNKubernetes","user_managed_networking":false,"cluster_networks":[{"cidr":"10.128.0.0/14","host_prefix":23}],"service_networks":[{"cidr":"172.30.0.0/16"}],"machine_networks":[{"cidr":"192.168.145.0/24"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
```

by running `watch -n1 curl ...` I see that for about 10 seconds API VIP is `null` and after this time it gets an IP from the new subnet.

I once again send the payload as above (i.e. the same machine network)

```
curl --data '{"vip_dhcp_allocation":true,"network_type":"OVNKubernetes","user_managed_networking":false,"cluster_networks":[{"cidr":"10.128.0.0/14","host_prefix":23}],"service_networks":[{"cidr":"172.30.0.0/16"}],"machine_networks":[{"cidr":"192.168.145.0/24"}]}' -H "content-type: application/json" --request PATCH http://chocobomb:6008/api/assisted-install/v1/clusters/82b8c284-8024-4a30-9bef-4b5a8180061b
```

and again by running `watch -n1 curl ...` I see that at no point API VIP was `null`.

## Assignees

/cc @YuviGold 
/cc @flaper87 

## Checklist


## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md
